### PR TITLE
Differential expression module

### PR DESCRIPTION
**Pull request by PierreBoyeau**

### CODE REVIEW

File: src/scvi_v2/_module.py [hunk 1]
```diff
@@ -249,3 +249,16 @@ def loss(
             reconstruction_loss=reconstruction_loss,
             kl_local=(kl_u + kl_z),
         )
+
+    def compute_h_from_x(self, x, sample_index, batch_index, cf_sample=None, continuous_covs=None, mc_samples=10):
+        """Compute normalized gene expression from observations"""
+        library = 7.0 * jnp.ones_like(sample_index)
```
Comment (justjhong) [inline src/scvi_v2/_module.py]: why is this 7.0?

File: src/scvi_v2/_model.py [hunk 2]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+        To correct for batch effects, the current approach consists in
+        computing batch-specific LFCs, and then averaging them.
+        Right now, other covariates are not considered.
+
+        Parameters
+        ----------
+        adata
+            AnnData object to use.
+        samples_a
+            Set of samples to use as reference.
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: is this reference or query or just first and second set of samples? The statistic doesn't seem directional and this description implies that the order matters

File: src/scvi_v2/_model.py [hunk 3]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+        samples_b
+            Set of samples to use as query.
+        """
+        mc_samples_per_obs = np.minimum((mc_samples_total // adata.n_obs), 1)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: this should be np.max?

File: src/scvi_v2/_model.py [hunk 4]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+        @jax.jit
+        def get_hs(inputs, cf_sample):
+            _h_inference_fn = lambda cf_sample: h_inference_fn(
+                inputs["x"],
+                inputs["sample_index"],
+                inputs["batch_index"],
+                cf_sample,
+                inputs["continuous_covs"],
+            )
+            hs = jax.lax.map(_h_inference_fn, cf_sample)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: vmap option?

File: src/scvi_v2/_model.py [hunk 5]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+        Right now, other covariates are not considered.
+
+        Parameters
+        ----------
+        adata
+            AnnData object to use.
+        samples_a
+            Set of samples to use as reference.
+        samples_b
+            Set of samples to use as query.
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: can you add documentation for all args here too

File: src/scvi_v2/_model.py [hunk 6]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+        self,
+        adata: AnnData,
+        samples_a: List[str],
+        samples_b: List[str],
+        delta: float = 0.5,
+        return_dist: bool = False,
+        batch_size: int = 128,
+        mc_samples_total: int = 10000,
+        max_mc_samples_per_pass: int = 50,
+        eps: float = 1e-4,
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: nit: reorder eps to the end

File: src/scvi_v2/_model.py [hunk 7]
```diff
@@ -537,3 +537,153 @@ def _get_scores(w, x, statistic):
...
+                inputs["sample_index"],
+                inputs["batch_index"],
+                cf_sample,
+                inputs["continuous_covs"],
+            )
+            hs = jax.lax.map(_h_inference_fn, cf_sample)
+
+            # convert to pseudocounts
+            hs = hs + eps
+            hs = jnp.log(hs)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]:
```suggestion
            hs = jnp.log(hs+eps)
```

**Review by justjhong — COMMENTED** [review]
looks great, left a few comments and added some asserts to the tests!

File: src/scvi_v2/_module.py (hunk 1 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_module.py]: I took 7.0 arbitrary, as it corresponds to a library size of exp(7) ~~ 1,000.
This is a placeholder and should not affect the value of the normalized expression levels.
We could use a value of zero for instance.
However, I was afraid that doing so we could have weird numerical issues when computing `h` as 
`h = px.mean / library_exp`.

**Review by justjhong — APPROVED** [review]
LGTM
